### PR TITLE
Onboard multi volume support

### LINKED ISSUE #2215: [FR] MULTI_VOLUME for onboard SD & onboard USB Support
It will be interesting to add support for multi volume to choose which volume (SD or USB) to mount.

Usefull for new boards like SKR 2, GTR, Octopus etc

It added to Marlin menu with this PR : https://github.com/MarlinFirmware/Marlin/pull/22004

Multi volume are working in Marlin mode but not on TFT mode.
In Marlin mode print menu gives access to the ability to browse onboard SD storage and onboard USB storage on the motherboard.
In TFT mode we have only access to the onboard USB storage with OnboardSD icon but not to SD storage when MULTI_VOLUME is enabled in Marlin.

### PR DESCRIPTION
**Pull request by kisslorand**
### Requirements

 - BTT or MKS TFT
 - motherboard with an USB storage port
 - Marlin (not older than 28.II.2022)

### Description

This PR adds the ability to select USB media also from motherboards which are equipped with it

### Benefits

Everyone using Marlin with motherboards that also have USB media storage port.

### Related Issues

Fixes #2215

### Notes

This PR is marked as draft because in Marlin there's no gcode for setting/switching the storage yet. However there's a pending PR for that on Marlin's Github repository (https://github.com/MarlinFirmware/Marlin/pull/23780).
For those who still want to use this feature here's what is needed to be done in Marlin source code in order to be able to use this feature:
 #### 1. in "M21_M22.cpp" replace
 ```
void GcodeSuite::M21() { card.mount(); }
```
   with
```
void GcodeSuite::M21()
{
  #if ENABLED(MULTI_VOLUME)
    if (parser.seen('S')) {  // "S" for SD Card
      card.changeMedia(&card.media_driver_sdcard);
    }
    else if (parser.seen('U')) {  // "U" for USB
      card.changeMedia(&card.media_driver_usbFlash);
    }
  #endif
  card.mount();
}
```

#### 2. in "M155.cpp" after the lines
```
    // SDCARD (M20, M23, M24, etc.)
    cap_line(F("SDCARD"), ENABLED(SDSUPPORT));
```
   add the folowing:
```
    // MULTI_VOLUME (M21 S/M21 U)
    #if ENABLED(SDSUPPORT)
      cap_line(F("MULTI_VOLUME"), ENABLED(MULTI_VOLUME));
    #endif
```

Enjoy !!!

L.E.: https://github.com/MarlinFirmware/Marlin/pull/23780 has been merged so this PR is fully functional without any Marlin modification.

### CODE REVIEW

File: TFT/src/User/API/parseACK.c [hunk 1]
```diff
@@ -1179,6 +1179,10 @@ void parseACK(void)
       {
         infoMachineSettings.onboardSD = ack_value();
       }
+      else if (ack_seen("Cap:MULTI_VOLUME:"))
```
Comment (digant73) [inline TFT/src/User/API/parseACK.c:1182]: as for Cap:SDCARD, a sort of check on infoSettings.onboard_sd could be done. Maybe onboard_sd renamed to onboard_media

File: Copy to SD Card root directory to update/config.ini [hunk 2]
```diff
@@ -461,8 +461,8 @@ ext_speed:S60 N600 F1200
 #   Options: [disable: 0, enable: 1]
 auto_load_leveling:1
 
-#### Onboard / Printer SD Card Support
-# On Marlin firmware, the TFT will auto-detect Onboard SD Card.
+#### Onboard / Printer Media Support
```
Comment (digant73) [inline Copy to SD Card root directory to update/config.ini:464]: the same changes should be applied to the other two config.ini in the project

File: TFT/src/User/API/Language/Language.inc [hunk 3]
```diff
@@ -180,6 +180,7 @@ X_WORD (USB_DISK_INSERTED)
 X_WORD (USB_DISK_REMOVED)
 X_WORD (USB_DISK_NOT_DETECTED)
 X_WORD (ONBOARDSD)
+X_WORD (ONBOARDUSB)
```
Comment (digant73) [inline TFT/src/User/API/Language/Language.inc:183]: same as my comment for language

File: TFT/src/User/API/PowerFailed.c [hunk 4]
```diff
@@ -52,7 +52,7 @@ bool powerFailedCreate(char *path)
 
   create_ok = false;
   if (!infoSettings.plr) return false;            // if disabled plr
-  if (infoFile.source >= BOARD_SD) return false;  // on board SD not supported now
+  if (infoFile.source >= BOARD_MEDIA) return false;  // on board SD not supported now
```
Comment (digant73) [inline TFT/src/User/API/PowerFailed.c:55]: apply also to the comment

File: Copy to SD Card root directory to update/Language Packs/language_am.ini [hunk 5]
```diff
@@ -154,6 +154,7 @@ label_usb_disk_inserted:U Disk inserted!
 label_usb_disk_removed:U Disk removed!
 label_usb_disk_not_detected:No USB Disk detected.
 label_onboardsd:OnboardSD
+label_onboardusb:OnboardUSB
```
[the remark this answers is not in the document]
Comment (kisslorand) [reply Copy to SD Card root directory to update/Language Packs/language_am.ini:157]: The report is the same. I just forgot to change "Read Onboard SD card error!" to "Read onboard media error!" in the English language file.

File: TFT/src/User/API/parseACK.c (hunk 1 above)
Comment (kisslorand) [reply TFT/src/User/API/parseACK.c:1182]: Why? What would it prevent? Who would benefit from this check?

> as for Cap:SDCARD, a sort of check on infoSettings.onboard_sd could be done. 

Why? What would it prevent? Who would benefit from this check?

> Maybe onboard_sd renamed to onboard_media

It's debatable... You have to enable onboard SD to be able to reach onboard USB. There's no way to have USB enabled without the SD enabled.
Anyway, Marlin refers to onboard USD also as "SD Card". Check the message Marlin gives when USB is mounted.
So let's keep the naming synced.

File: Copy to SD Card root directory to update/config.ini (hunk 2 above)
Comment (kisslorand) [reply Copy to SD Card root directory to update/config.ini:464]: Roger that!

File: TFT/src/User/API/Language/Language.inc (hunk 3 above)
Comment (kisslorand) [reply TFT/src/User/API/Language/Language.inc:183]: Noted

File: TFT/src/User/API/PowerFailed.c (hunk 4 above)
Comment (kisslorand) [reply TFT/src/User/API/PowerFailed.c]: Done

File: TFT/src/User/API/Gcode/gcode.c [hunk 6]
```diff
@@ -73,7 +73,7 @@ bool request_M21(void)
                           "SD init fail",         // The second error magic
                           "volume.init failed");  // The third error magic
 
-  mustStoreCmd("M21\n");
+    mustStoreCmd((infoMachineSettings.multiVolume == ENABLED) ? ((infoFile.boardSource == BOARD_SD) ? "M21 S\n" : "M21 U\n") : "M21\n");
```
Comment (digant73) [inline TFT/src/User/API/Gcode/gcode.c:76]: wrong indentation

File: TFT/src/User/API/Language/language_jp.h [hunk 7]
```diff
@@ -173,7 +173,8 @@
     #define STRING_USB_DISK_REMOVED       "U Diskカード出しました!"
     #define STRING_USB_DISK_NOT_DETECTED  "No USB Disk detected."
     #define STRING_ONBOARDSD              "内部 mSDカード"
-    #define STRING_READ_ONBOARDSD_ERROR   "内部 mSDカードの読み取りがエラー発生しました!"
+     #define STRING_ONBOARDUSB             "OnboardUSB"
```
Comment (digant73) [inline TFT/src/User/API/Language/language_jp.h:176]: two lines with wrong indentation

File: README.md [hunk 8]
```diff
@@ -99,7 +99,7 @@ In order the TFT firmware is able to provide all of its functionalities/features
 `M114_DETAIL` (in Configuration_adv.h)<br>
 `REPORT_FAN_CHANGE` (in Configuration_adv.h)<br>
 
-**Options to support printing from onboard SD:**
+**Options to support printing from onboard media:**
```
Comment (digant73) [inline README.md:102]: you should apply that change also on top of all the 3 config.ini files

Comment (kisslorand) [reply README.md:102]: I am not sure what do you suggest here. Can you be please more specific?

File: TFT/src/User/API/Gcode/gcode.c (hunk 6 above)
Comment (kisslorand) [reply TFT/src/User/API/Gcode/gcode.c:76]: Fixed.

File: TFT/src/User/API/Language/language_jp.h (hunk 7 above)
Comment (kisslorand) [reply TFT/src/User/API/Language/language_jp.h:176]: Fixed.

File: README.md (hunk 8 above)
Comment (digant73) [reply README.md:102]: the same text is provided at the beginning of each config.ini file. So also that text should be modified:

`# Options to support printing from onboard SD:
`
to

`# Options to support printing from onboard media:
`

Comment (kisslorand) [reply README.md:102]: I cannot find "# Options to support printing from onboard SD:" left unchanged.

Comment (digant73) [reply README.md:102]: oh sorry, my mistake. I didn't see that change

Comment (kisslorand) [reply README.md:102]: No problem, I thought my VSCode is messing with me :)

**Comment by kisslorand** [comment]
Thank you for the feedback.

**Comment by kisslorand** [comment]
@bigtreetech 
https://github.com/MarlinFirmware/Marlin/pull/23780 has been merged so this PR is fully functional without any Marlin modification, it can be merged.

**Pull request merged**